src/hardware/input/intel8042.cpp

```diff
@@ -0,0 +1,1044 @@
+}
+
+static void request_system_reset()
+{
+	E_Exit("I8042: System reset requested");
+	// TODO: we need a proper reset implementation
+}
+
+static void execute_command(const Command command)
+{
+	// LOG_INFO("I8042: Command 0x%02x", static_cast<int>(command));
+
+	switch (command) {
+	//
+	// Commands requiring a parameter
+	//
+	case Command::WriteByteConfig:  // 0x60
+	case Command::WriteOutputPort:  // 0xd1
+	case Command::SimulateInputKbd: // 0xd2
+	case Command::SimulateInputAux: // 0xd3
+	case Command::WriteAux:         // 0xd4
+		current_command = command;
+		break;
+	case Command::WriteControllerMode: // 0xcb
+		warn_controller_mode();
+		current_command = command;
+		break;
+
+	//
+	// No-parameter commands
+	//
+	case Command::ReadByteConfig: // 0x20
+		// Reads the keyboard controller configuration byte
+		buffer_add(config_byte.data);
+		break;
+	case Command::ReadFwCopyright: // 0xa0
+		// Reads the keyboard controller firmware
+		// copyright string, terminated by NUL
+		for (auto byte : firmware_copyright) {
+			buffer_add(static_cast<uint8_t>(byte));
+		}
+		buffer_add(0);
+		break;
+	case Command::ReadFwRevision: // 0xa1
+		// Reads the keybaord copntroller firmware
+		// revision, always one byte
+		buffer_add(firmware_revision);
+		break;
+	case Command::PasswordCheck: // 0xa4
+		// Check if password installed
+		// 0xf1: not installed, or no hardware support
+		// 0xfa: password installed
+		buffer_add(0xf1);
+		break;
+	case Command::DisablePortAux: // 0xa7
+		// Disable aux (mouse) port
+		is_disabled_aux = true;
+		break;
+	case Command::EnablePortAux: // 0xa8
+		// Enable aux (mouse) port
+		is_disabled_aux = false;
+		break;
+	case Command::TestPortAux: // 0xa9
+		// Port test. Possible results:
+		// 0x01: clock line stuck low
+		// 0x02: clock line stuck high
+		// 0x03: data line stuck low
+		// 0x04: data line stuck high
+		// Disables the aux (mouse) port
+		is_disabled_aux = true;
+		buffer_add(0x00);
+		break;
+	case Command::TestController: // 0xaa
+		// Controller test. Possible results:
+		// 0x55: passed; 0xfc: failed
+		// Disables aux (mouse) and keyboard ports
+		is_disabled_aux = true;
+		is_disabled_kbd = true;
+		buffer_add(0x55);
+		break;
+	case Command::TestPortKbd: // 0xab
+		// Port test. Possible results:
+		// (as with aux port test)
+		// Disables the keyboard port
+		is_disabled_kbd = true;
+		buffer_add(0x00); // as with TestPortAux
+		break;
+	case Command::DiagnosticDump: // 0xac
+		// Dump the whole controller internal RAM (16 bytes),
+		// output port, input port, and status port
+		warn_internal_ram_access();
+		buffer_add(config_byte.data);
+		for (uint8_t idx = 1; idx <= 16; idx++) {
+			buffer_add(0);
+		}
+		buffer_add(get_input_port());
+		buffer_add(get_output_port());
+		buffer_add(status_byte.data);
+		break;
+	case Command::DisablePortKbd: // 0xad
+		// Disable keyboard port; any keyboard command
+		// reenables the port
+		is_disabled_kbd = true;
+		break;
```

Comment:
This is extremely readable and self-documenting. It doesn't get better than this!